akka-actor/src/main/scala-2.13/akka/util/ByteString.scala

```diff
@@ -271,6 +269,11 @@ object ByteString {
       toCopy
     }
 
+    override def toArrayUnsafe(): Array[Byte] = bytes
+
+    override def foldLeftArrayUnsafe[T](zero: T)(f: (T, Array[Byte], Int, Int) => T): T =
```

Comment:
Ok, this should be possible to use with `proto.UnsafeByteOperations.unsafeWrap(bytes, offset, length), and `proto.ByteString.concat`.

An alternative could be to use the existing `akka.ByteString.asByteBuffers.c`.

Is there a public api to see if it's a ByteString1? Because then we could use that to choose between toArrayUnsafe or slightly more costly toArrayUnsafe.

Reply:
The concrete ByteString classes are actually public, so you could use `isInstanceOf` to detect `ByteString1`.